### DISCUSSION TITLE
Return state data without cloning 🤔

### DISCUSSION THREAD

**Discussion by NiiightmareXD** [tokio-rs/axum#2832] [Q&A] [answered]
### Summary

I'm building an experimental cache system to reduce database calls for the API. However, I'm facing a challenge: cloning the T data or Vec<T> is necessary because the handler can't return an Arc. Are there any workarounds to avoid cloning?

this is my current implementation:

```rs
#[derive(Clone)]
pub struct CacheHandler {
    connection: ConnectionManager,
    posts_cache: Arc<RwLock<Option<Vec<Post>>>>,
    sliders_cache: Arc<RwLock<Option<Sliders>>>,
}

impl CacheHandler {
    pub async fn posts(&mut self) -> ServerResult<Vec<Post>> {
        let posts = self.posts_cache.read().await.clone();
        if let Some(posts) = posts {
            Ok(posts)
        } else {
            let mut keys: Vec<String> = self.connection.keys("post:*").await?;
            keys.sort_unstable();

            let mut posts: Vec<Post> = Vec::new();
            for key in keys {
                let post: Post = self.connection.get(key).await?;

                posts.push(post);
            }

            *self.posts_cache.write().await = Some(posts.clone());

            Ok(posts)
        }
    }

    pub async fn sliders(&mut self) -> ServerResult<Sliders> {
        let sliders = self.sliders_cache.read().await.clone();
        if let Some(sliders) = sliders {
            Ok(sliders)
        } else {
            let sliders: Sliders = self.connection.get("sliders").await?;

            *self.sliders_cache.write().await = Some(sliders.clone());

            Ok(sliders)
        }
    }
}
```

Ideally, the handler would return Arc.

### axum version

0.7.5

**Comment by jplatte** [comment]
What's the actual type you would want to return from your handler(s)?

**Reply by NiiightmareXD** [reply]
> What's the actual type you would want to return from your handler(s)?

It's `Vec<Post>` which has a few `String` in it.

**Reply by jplatte** [reply]
I'm a bit confused, `Vec<u8>` implements `IntoResponse` but no other `Vec<_>` does. Do you want to JSON-encode it, or something else?

**Reply by NiiightmareXD** [reply]
Yes, it's wrapped in a Json<T> but if I do Json<Arc<T>> it doesn't work which makes it so that I have to clone every time which I think is not the correct approach and I'm missing something.

**Reply by jplatte** [reply, marked as answer]
You can actually do `Json<Arc<_>>` if you enable serde's `rc` feature: https://docs.rs/serde/latest/serde/trait.Serialize.html#impl-Serialize-for-Arc%3CT%3E

Alternatively, you can use axum-extra's [`ErasedJson`](https://docs.rs/axum-extra/0.9/axum_extra/response/struct.ErasedJson.html) type and pass a reference to the arc-wrapped value to its constructor.

Finally, we could probably make axum support `Arc<Json<_>>` as a response type, but I'm not sure how useful that would be here.

**Reply by NiiightmareXD** [reply]
Thanks for the help I'm pretty sure that would be the most performant way of doing this but also after seeing the docs I realized serve also implements it for [Cow<'a, T>](https://doc.rust-lang.org/nightly/alloc/borrow/enum.Cow.html) I'm not sure which one would be better.

**Reply by jplatte** [reply]
`Cow` is not going to help you here. You can't return `Cow<'a, T>` from a handler function (unless you count `'a` being `'static` but then you could also return `Json<&'static Vec<_>>`). I would generally recommend the `ErasedJson` solution, since `serde`s `rc` is off by default for a good reason (it can be somewhat of a footgun when used for deserialization).

**Reply by NiiightmareXD** [reply]
> `Cow` is not going to help you here. You can't return `Cow<'a, T>` from a handler function (unless you count `'a` being `'static` but then you could also return `Json<&'static Vec<_>>`). I would generally recommend the `ErasedJson` solution, since `serde`s `rc` is off by default for a good reason (it can be somewhat of a footgun when used for deserialization).

Yea seems like it will lose its identity after deserialization it doesn't matter here though
